### PR TITLE
Sts generic provider

### PR DESCRIPTION
**Pull request by PhuongLe**
There are 3 things have been done
1. Update Safewhere.Samples.STS.Common 
+ To add more binding for calling to username endpoint 
+ Improve configuration to make it able to call to username endpoint and define additional claims added to RST

2. Add project Safewhere.Samples.STS.GenericProviderAppConsole
+ The project is a console application which will negotiate token from identify by using username endpoint.
+ The RST has some additional predefined claims

3. Add project Safewhere.Samples.STS.GenericCredentialsValidator
+ This is to write a sample of generic credential validator
+ This sample is quite simple. It will validate if there are enough information or not. If yes, then write successful message on debug log file. Otherwise, return false.
+ The expected values on settings are: username, password, serviceidentifier, [optional claims]

### CODE REVIEW

File: Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                KeyType = KeyTypes.Symmetric,
+                Issuer = new EndpointReference(rstConfiguration.StsEndpointAddress.Uri.AbsoluteUri),
+            };
+            if (bootstrapToken != null)
+            {
+                certificateRst.ActAs = new SecurityTokenElement(bootstrapToken);
+            }
+            if (rstConfiguration.ClientCertificate == null)
+            {
+                throw new IdentifyStsProcessException("Cannot execute negotiating token request to certificate endpoint withou client cerificate");
```
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs]: withou -> without

File: Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            try
+            {
+                var stschannel = CreateStsChannel(rstConfiguration.ClientCertificate, rstConfiguration.StsEndpointAddress);
+                return stschannel.Issue(certificateRst);
+            }
+            catch (Exception ex)
+            {
+                Logging.Instance.Error(ex, "There is an error responsed from ws-trust service. The test will be failed");
```
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs]: will be failed -> wrong grammar

File: Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs (hunk 1 above)
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs]: You need to check other places for this same mistake too

File: Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs (hunk 2 above)
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs]: You need to check other places for this same mistake too

File: Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                logWriter.WriteError("Validate the input failed. Error code: " + validationResult.ResultCode +
+                                       ". ExternalErrorMessages: " + string.Join("\n-", validationResult.ExternalErrorMessages));
+                validationResult.ShowErrorViewWhenResultCodeIsNotSuccess = false;
+                return validationResult;
+            }
+            var username = inputs[UserName];
+            var serviceIdentifier = inputs[ServiceIdentifier];
+            var logResult = new StringBuilder();
+            logResult.AppendLine(
+                $"Successfully validate generic credentials for username = '{username}' with service identifier ='{serviceIdentifier}'");
```
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs]: Nice use of new syntax. Hopefully our clients are using VS2015 ;))

File: Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            logWriter.WriteInformation(logResult); 
+
+            return new CredentialsValidationResult {ResultCode = CredentialsValidationResultCode.Success};
+        }
+        private static CredentialsValidationResult ValidateInput(IDictionary<string, string> inputs)
+        {
+            CredentialsValidationResult validationResult = new CredentialsValidationResult();
+            if (!inputs.ContainsKey(ServiceIdentifier))
+            {
+                validationResult.ExternalErrorMessages.Add("Missing connection string field");
```
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs]: There is no "Connection string" here :D

File: Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            foreach (var input in inputs)
+            {
+                if (input.Key.StartsWith(AdditionalClaims, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    logResult.AppendLine($"Additional claims received: type = '{input.Key}' - value ='{input.Value}'");
+                }
+            }
+            logWriter.WriteInformation(logResult); 
+
+            return new CredentialsValidationResult {ResultCode = CredentialsValidationResultCode.Success};
```
Comment (thuannguy) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs]: I think you should demonstrate a failed case. For example, if username is "demowrongusername" and password is "demowrongpassword", this method will return a fail result.

File: Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs (hunk 1 above)
Comment (PhuongLe) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.Common/IdentifyStsCertificateEndpointHandler.cs]: fixed, only 1 mistake like it

File: Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs (hunk 4 above)
Comment (PhuongLe) [inline Safewhere.Samples.STS/Safewhere.Samples.STS.GenericCredentialsValidator/UsernamePasswordGenericCredentialsValidator.cs]: oops, just fixed it

**Pull request merged**